hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientResult.java

```diff
@@ -49,6 +50,7 @@
 
     private boolean closed;
     private boolean iteratorAccessed;
+    private ClientInvocationFuture pendingPageFuture;
```

Comment:
I vote for avoiding this asynchrony now. It complicates things while at the same time it is not proven that it brings anything in real use cases. The current implementation has two issues:
1. It lacks synchronization on `close/cancel`. One of the typical use cases is closing the cursor from the *other* thread, not the same. Even JDBC standard assumes such usage pattern
1. We do not want to have request-response flow at all. In the mid-term, we are going to rework client communication to streaming mode, when multiple responses are sent from the server to the client. This will make the current implementation obsolete. And it is likely to happen very soon.

Reply:
I don't agree, but i'm removing this part. It's kind of unrelated to the main issue. Btw, the rest of the `close` method also isn't thread safe.

The benefit is: you don't initiate the round trip after you're done with the previous batch, but you initiate it as you start processing it. Clients with longer RTT and doing non-trivial processing with the rows will benefit the most. Having it will not prevent us to switch to streaming mode in the future.

Reply:
Yes, there definitely benefits for non-trivial processing. And perhaps it is more prevalent in streaming, than in IMDG cases. 
The "streaming" client communication mode I mentioned is likely to work very differently even comparing to the current implementation in master. As the client team explained to me, the protocol is designed in a way to allow for sending chunks of data one after the other from the server without an explicit request from the client. So it is likely that the final implementation of the protocol will send just one request to the server, and then read chunks continuously with some sort of backpressure. 

It is likely that we will implement that soon because this is a breaking change, and we cannot be in the beta forever.

Reply:
Yeah, but why not do simple improvements to the current version?...

Reply:
My concern is that it is not well aligned with the planned protocol and JDBC changes, so it will increase the design and implementation burden. 
Maybe let's merge it to `master` as is, and prepare the PR for this change separately, and then listen for client team opinion? This way we will merge the main part, where we have a consensus, quickly.